src/braket/experimental/autoqasm/api.py

```diff
@@ -114,40 +127,50 @@ def gate_calibration(*args, implements: Callable, **kwargs) -> Callable[[], Gate
         `with_calibrations` method of the main program.
     """
     return _function_wrapper(
-        *args,
+        None,
         converter_callback=_convert_calibration,
         converter_args={"gate_function": implements, **kwargs},
     )
 
 
 def _function_wrapper(
-    *args: tuple[Any],
+    func: Optional[Callable],
+    *,
     converter_callback: Callable,
     converter_args: Optional[dict[str, Any]] = None,
-) -> Callable[[Any], aq_program.Program]:
+) -> Callable[..., Optional[Union[aq_program.Program, GateCalibration]]]:
```

Comment:
This return type hint (as for all of the functions above which call it) is still not correct in the case where we `return functools.partial(...)`. Can we make recursive type hints which `_function_wrapper` and all of the decorators can use for their return type hints? Haven't tried it, but I guess something like (maybe this can be simplified):
```
ProgramDecoratorRetVal = Callable[...,   # for main(), subroutine()
    Union['ProgramDecoratorRetVal', aq_program.Program]]
GateDecoratorRetVal = Callable[...,   # for gate()
    Optional['GateDecoratorRetVal']]
GateCalibrationDecoratorRetVal = Callable[...,    # for gate_calibration()
    Union['GateCalibrationDecoratorRetVal', GateCalibration]]
DecoratorRetVal = Union[    # for _function_wrapper()
    ProgramDecoratorRetVal, GateDecoratorRetVal, GateCalibrationDecoratorRetVal]
```

Reply:
aside: wish we could write `A | B | None` instead of `Optional[Union[A, B]]`!

Reply:
Yes, I was thinking to update the type hints in another PR. We can use `|` syntax if we import `__annotations__` from future